code/client/src/components/DomainPurchaseModal.jsx

```diff
@@ -0,0 +1,222 @@
+
+const { Text, Title } = Typography
+
+const inputStyle = {
+  borderWidth: '0 0 2px',
+  borderStyle: 'dashed',
+  borderColor: '#000000',
+  padding: 0,
+  textAlign: 'right'
+}
+
+const priceRowStyle = {
+  textAlign: 'center'
+}
+
+const inputRowStyle = {
+  paddingBottom: '30px'
+}
+
+const WarningTextStyle = {
+  textAlign: 'center',
+  marginBottom: '20px',
+  display: 'block'
+}
+
+const minDomainNameLength = 3
+
+const delayCheckMillis = 1300
+
+const oneDomain = '.crazy.one'
+
+const validDomain = (domainName) => {
+  try {
+    if (domainName.length <= minDomainNameLength) {
+      return undefined
+    }
+
+    return ONEUtil.normalizeDomain(`${domainName}${oneDomain}`)
+  } catch (e) {
+    return undefined
+  }
+}
+
+/**
+ * Custom hook that executes a function with delay and cancellation, if the useEffect is destroyed due to the dependencies
+ * update, the timeout is cancelled, which cancels the function execution.
+ * The function only runs when the supplied condition is true.
+ */
+const useWaitExecution = (func, runCondition, wait, dependencies) => {
+  useEffect(() => {
+    let timeout
+    if (runCondition) {
+      timeout = setTimeout(func, wait)
+    }
+
+    return () => {
+      clearTimeout(timeout)
+    }
+  }, dependencies)
+}
+
+/**
+ * Renders warning message block for the ability to purchase a domain based on the domain availability and balance availability.
+ */
+const WarningMessageBlock = ({ enoughBalance, domainAvailable, checkingAvailability, validatedDomain }) => (
+  <Space direction='vertical' style={WarningTextStyle}>
+    {
+      !enoughBalance && !checkingAvailability ? <Warning>Not enough ONE balance</Warning> : <></>
+    }
+    {
+      !domainAvailable && !checkingAvailability ? <Warning>Domain is not available</Warning> : <></>
+    }
+    {
+      checkingAvailability && validatedDomain ? <Spin /> : <></>
+    }
+  </Space>
+)
+
+/**
+ * Renders a modal that enables users to purchase an available domain for their selected wallet using selected token.
+ */
+const DomainPurchaseModal = ({ isModalVisible, dismissModal, oneBalance, walletAddress }) => {
+  const dispatch = useDispatch()
+
+  const [domainName, setDomainName] = useState('')
+
+  const [purchaseOnePrice, setPurchaseOnePrice] = useState(0)
+
+  const [domainFiatPrice, setDomainFiatPrice] = useState(0)
+
+  const [available, setAvailable] = useState(false)
+
+  const [enoughBalance, setEnoughBalance] = useState(false)
+
+  const [domainAvailable, setDomainAvailable] = useState(false)
+
+  const [checkingAvailability, setCheckingAvailability] = useState(true)
+
+  const price = useSelector(state => state.wallet.price)
+
+  const validatedDomain = validDomain(domainName)
+
+  const purchaseDomain = useCallback(async () => {
+    // The validated domain will be sent as [selectedDomainName].crazy.one.
+    dispatch(walletActions.purchaseDomain({ domainName: validatedDomain, address: walletAddress }))
```

Comment:
@polymorpher I'm not sure what is expected to be sent with purchase domain request, selected domain or full domain with .crazy.one. If not correct, please let me know to update this.

Reply:
It will be a SmartFlows.commitReveal(...). I will add it. The dispatcher would not be a good place to call that since SmartFlows.commitReveal has lots of callbacks and each would read or modify the states of the component to compute for proofs and reflect the progress.